diff -rauN py2app-0.5.2/py2app/build_app.py patched-py2app/build_app.py
--- py2app-0.5.2/py2app/build_app.py	2010-07-26 00:05:08.000000000 +1000
+++ patched-py2app/build_app.py	2010-12-01 16:21:49.000000000 +1100
@@ -29,6 +29,7 @@
 from modulegraph.modulegraph import SourceModule, Package, os_listdir
 
 import macholib.dyld
+import macholib.util
 import macholib.MachOStandalone
 
 from py2app.create_appbundle import create_appbundle
@@ -603,6 +604,7 @@
             else:
                 self.run_normal()
         except:
+            raise
             # XXX - remove when not debugging
             #       distutils sucks
             import pdb, sys, traceback
@@ -775,6 +777,7 @@
             if self.semi_standalone:
                 self.symlink(sys.executable, execdst)
             else:
+                execsrc = sys.executable
                 if os.path.exists(os.path.join(sys.prefix, ".Python")):
                     fn = os.path.join(sys.prefix, "lib", "python%d.%d"%(sys.version_info[:2]), "orig-prefix.txt")
                     if os.path.exists(fn):
@@ -785,10 +788,10 @@
                         rest_path = rest_path[1:]
 
                     print "XXXX", os.path.join(prefix, rest_path)
-                    self.copy_file(os.path.join(prefix, rest_path), execdst)
+                    execsrc = os.path.join(prefix, rest_path)
 
-                else:
-                    self.copy_file(sys.executable, execdst)
+                self.copy_file(execsrc, execdst)
+                macholib.MachOStandalone.rewrite_loader_path(execdst,execsrc)
             if not self.debug_skip_macholib:
                 mm = PythonStandalone(self, dst, executable_path=exp)
                 dylib, runtime = self.get_runtime()
@@ -885,6 +888,7 @@
         if src != sys.executable:
             force, self.force = self.force, True
             self.copy_file(src, dst)
+            macholib.MachOStandalone.rewrite_loader_path(dst,src)
             self.force = force
         return dst
 
@@ -1275,6 +1279,7 @@
             )
             self.mkpath(os.path.dirname(fn))
             copy_file(copyext.filename, fn, dry_run=self.dry_run)
+            macholib.MachOStandalone.rewrite_loader_path(fn,copyext.filename)
 
         target.appdir = appdir
         return appdir
diff -rauN py2app-0.5.2/py2app/util.py patched-py2app/util.py
--- py2app-0.5.2/py2app/util.py	2010-07-14 00:45:01.000000000 +1000
+++ patched-py2app/util.py	2010-12-01 16:21:49.000000000 +1100
@@ -589,3 +589,7 @@
     MAPC = '/Developer/usr/bin/mapc'
 def mapc(src, dst):
     os.spawnv(os.P_WAIT, MAPC, [MAPC, src, dst])
+
+
+def rewrite_loader_path(newpath,oldpath):
+    return macholib.util.rewrite_loader_path(newpath,oldpath)
